private/buf/bufwire/proto_encoding_writer.go

```diff
@@ -60,14 +59,26 @@ func (i *protoEncodingWriter) PutMessage(
 	if err != nil {
 		return err
 	}
-	data, err := protoencoding.NewJSONMarshalerIndent(resolver).Marshal(message)
-	if err != nil {
-		return err
+	var marshaler protoencoding.Marshaler
+	switch messageRef.MessageEncoding() {
+	case bufconvert.MessageEncodingBin:
+		marshaler = protoencoding.NewWireMarshaler()
+	case bufconvert.MessageEncodingJSON:
+		marshaler = protoencoding.NewJSONMarshalerIndent(resolver)
+	default:
+		return fmt.Errorf("unknown message encoding type")
```

Comment:
Similar to above w.r.t. earlier validation, but this one is less of a big deal because there's only one other thing we're doing before this potential error.

Reply:
for this one, the resolver is actually need to be existed before this switch block, as we need it to create the marshaler, so the only way we can return the error earlier is to do an additional checking in the beginning to check the value of `messageRef.MessageEncoding()`, which doesn't seems to be worth to do as this should only happen in an implementation error, how do you think?